### PR TITLE
Change $context['contenttype'] to be the same as $content->getContentType()

### PR DESCRIPTION
**Pull request by graham73may**
While working on modified content using the `PRE_HYDRATE` and `POST_HYDRATE` events I noticed that although you are able to edit the `$content`, the content type is instantly overwritten within the `action` method at `bolt/src/Storage/ContentRequest/Edit.php::153`.

This request changes this behaviour so you are able to ad-hoc alter the content type values, allowing a full dynamic approach to things such as the `route_prefix`.

I came across this issue specifically when trying to build a hierarchical page structure using the composer autoloader and a new service provider.

Of course an even better option would be to add events to the dispatcher for when the code gets/sets config values (allowing complete dynamic alterations without necessarily saving them) but as I understand it the config is being rewritten so I didn't want to mess too much with it.

Oh, and that thing about brown M & Ms....

### CODE REVIEW

File: src/Storage/ContentRequest/Edit.php [hunk 1]
```diff
@@ -125,43 +134,46 @@ public function action(Content $content, array $contentType, $duplicate)
 
         // Test write access for uploadable fields.
         $contentType['fields'] = $this->setCanUpload($contentType['fields']);
-        $templateFields = $content->getTemplatefields();
-        if ($templateFields instanceof TemplateFields && $templateFieldsData = $templateFields->getContenttype()->getFields()) {
-            $templateFields->getContenttype()['fields'] = $this->setCanUpload($templateFields->getContenttype()->getFields());
+        $templateFields        = $content->getTemplatefields();
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 2]
```diff
@@ -204,6 +217,7 @@ private function getRelationsList(array $contentType)
      */
     private function neededFields($relationValues, $relationConfig)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 3]
```diff
@@ -217,7 +231,6 @@ private function neededFields($relationValues, $relationConfig)
 
     /**
      * Determine write access for upload fields, and auto-create the desired directory if it does not exist.
-     *
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Re-add

File: src/Storage/ContentRequest/Edit.php [hunk 4]
```diff
@@ -227,6 +240,7 @@ private function neededFields($relationValues, $relationConfig)
      */
     private function setCanUpload($fields)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 5]
```diff
@@ -251,6 +265,7 @@ private function setCanUpload($fields)
      */
     private function checkUploadDirectory($path)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 6]
```diff
@@ -269,15 +284,16 @@ private function checkUploadDirectory($path)
...
-     * @param Content $content
+     * @param ContentType $contentType
+     * @param Content     $content
      *
      * @return array
      */
-    private function getTemplateFieldTemplates(array $contentType, Content $content)
+    private function getTemplateFieldTemplates(ContentType $contentType, Content $content)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 7]
```diff
@@ -269,15 +284,16 @@ private function checkUploadDirectory($path)
...
         $templateFieldTemplates = [];
-        $templateFieldsConfig = $this->config->get('theme/templatefields');
+        $templateFieldsConfig   = $this->config->get('theme/templatefields');
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 8]
```diff
@@ -311,6 +327,7 @@ private function getTemplateFieldTemplates(array $contentType, Content $content)
      */
     private function getPublishingDate($date, $setNowOnEmpty = false)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 9]
```diff
@@ -323,19 +340,22 @@ private function getPublishingDate($date, $setNowOnEmpty = false)
...
+     * @param ContentType $contentType
+     * @param array       $has
      *
      * @return array
      */
-    private function createGroupTabs(array $contentType, array $has)
+    private function createGroupTabs(ContentType $contentType, array $has)
     {
-        $groups = [];
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 10]
```diff
@@ -323,19 +340,22 @@ private function getPublishingDate($date, $setNowOnEmpty = false)
...
+        $groups   = [];
         $groupIds = [];
 
         $addGroup = function ($group, $label) use (&$groups, &$groupIds) {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 11]
```diff
@@ -323,19 +340,22 @@ private function getPublishingDate($date, $setNowOnEmpty = false)
...
             $nr = count($groups) + 1;
-            $id = rtrim('tab-' . Slugify::create()->slugify($group), '-');
+            $id = rtrim('tab-' . Slugify::create()
+                                        ->slugify($group), '-');
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: One line

File: src/Storage/ContentRequest/Edit.php [hunk 12]
```diff
@@ -345,15 +365,20 @@ private function createGroupTabs(array $contentType, array $has)
                 'is_active' => $nr === 1,
                 'fields'    => [],
             ];
-            $groupIds[$id] = 1;
+            $groupIds[$id]  = 1;
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 13]
```diff
@@ -345,15 +365,20 @@ private function createGroupTabs(array $contentType, array $has)
...
         };
 
         foreach ($contentType['groups'] ? $contentType['groups'] : ['ungrouped'] as $group) {
             if ($group === 'ungrouped') {
                 $addGroup($group, Trans::__('contenttypes.generic.group.ungrouped'));
             } elseif ($group !== 'meta' && $group !== 'relations' && $group !== 'taxonomy') {
                 $default = ['DEFAULT' => ucfirst($group)];
-                $key = ['contenttypes', $contentType['slug'], 'group', $group];
+                $key     = [
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 14]
```diff
@@ -387,25 +412,31 @@ private function createGroupTabs(array $contentType, array $has)
...
+     * @param ContentType $contentType
+     * @param Content     $content
+     * @param array       $has
      *
      * @return array
      */
-    private function getUsedFieldtypes(array $contentType, Content $content, array $has)
+    private function getUsedFieldtypes(ContentType $contentType, Content $content, array $has)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 15]
```diff
@@ -387,25 +412,31 @@ private function createGroupTabs(array $contentType, array $has)
...
+                                      ->getFields() ?: [];
         } else {
             $templateFields = [];
         }
 
-        foreach ([$contentType['fields'], $templateFields] as $fields) {
+        foreach ([
+                     $contentType['fields'],
+                     $templateFields
+                 ] as $fields) {
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: One line, or an assignment

File: src/Storage/ContentRequest/Edit.php [hunk 16]
```diff
@@ -125,43 +134,46 @@ public function action(Content $content, array $contentType, $duplicate)
...
+                                                                                             ->getFields());
         }
 
         // Build context for Twig.
-        $contextCan = [
-            'upload'             => $this->users->isAllowed('files:uploads'),
-            'publish'            => $this->users->isAllowed('contenttype:' . $contentTypeSlug . ':publish:' . $content->getId()),
-            'depublish'          => $this->users->isAllowed('contenttype:' . $contentTypeSlug . ':depublish:' . $content->getId()),
-            'change_ownership'   => $this->users->isAllowed('contenttype:' . $contentTypeSlug . ':change-ownership:' . $content->getId()),
+        $contextCan    = [
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 17]
```diff
@@ -125,43 +134,46 @@ public function action(Content $content, array $contentType, $duplicate)
...
+            'upload'           => $this->users->isAllowed('files:uploads'),
+            'publish'          => $this->users->isAllowed('contenttype:' . $contentTypeSlug . ':publish:' . $content->getId()),
+            'depublish'        => $this->users->isAllowed('contenttype:' . $contentTypeSlug . ':depublish:' . $content->getId()),
+            'change_ownership' => $this->users->isAllowed('contenttype:' . $contentTypeSlug . ':change-ownership:' . $content->getId()),
         ];
-        $contextHas = [
+        $contextHas    = [
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 18]
```diff
@@ -125,43 +134,46 @@ public function action(Content $content, array $contentType, $duplicate)
...
-            'contentowner'       => $contentowner,
-            'fields'             => $this->config->fields->fields(),
-            'fieldtemplates'     => $this->getTemplateFieldTemplates($contentType, $content),
-            'fieldtypes'         => $this->getUsedFieldtypes($contentType, $content, $contextHas),
-            'groups'             => $this->createGroupTabs($contentType, $contextHas),
-            'can'                => $contextCan,
-            'has'                => $contextHas,
-            'values'             => $contextValues,
-            'relations_list'     => $this->getRelationsList($contentType),
+        $context       = [
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 19]
```diff
@@ -171,22 +183,23 @@ public function action(Content $content, array $contentType, $duplicate)
...
      *
-     * @param array $contentType
+     * @param ContentType $contentType
      *
      * @return array
      */
-    private function getRelationsList(array $contentType)
+    private function getRelationsList(ContentType $contentType)
     {
+
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Remove

File: src/Storage/ContentRequest/Edit.php [hunk 20]
```diff
@@ -171,22 +183,23 @@ public function action(Content $content, array $contentType, $duplicate)
...
         $list = [];
         if (!isset($contentType['relations']) || !is_array($contentType['relations'])) {
             return $list;
         }
 
         foreach ($contentType['relations'] as $relationName => $relationValues) {
             /** @var Repository\ContentRepository $repo */
-            $repo = $this->em->getRepository($relationName);
+            $repo           = $this->em->getRepository($relationName);
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 21]
```diff
@@ -171,22 +183,23 @@ public function action(Content $content, array $contentType, $duplicate)
...
             $relationConfig = $this->config->get('contenttypes/' . $relationName, []);
-            $neededFields = $this->neededFields($relationValues, $relationConfig);
+            $neededFields   = $this->neededFields($relationValues, $relationConfig);
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 22]
```diff
@@ -55,29 +56,36 @@ public function __construct(
...
      */
-    public function action(Content $content, array $contentType, $duplicate)
+    public function action(Content $content, ContentType $contentType, $duplicate)
     {
+
         $contentTypeSlug = $contentType['slug'];
-        $new = $content->getId() === null ?: false;
-        $oldStatus = $content->getStatus();
-        $allStatuses = ['published', 'held', 'draft', 'timed'];
+        $new             = $content->getId() === null ?: false;
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 23]
```diff
@@ -55,29 +56,36 @@ public function __construct(
...
+        $oldStatus       = $content->getStatus();
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

File: src/Storage/ContentRequest/Edit.php [hunk 24]
```diff
@@ -55,29 +56,36 @@ public function __construct(
...
+        $allStatuses     = [
```
Comment (GawainLynch) [inline src/Storage/ContentRequest/Edit.php]: Revert spacing

**Comment by graham73may** [comment]
Sorry about that, IDE got a bit creative with the code style there!

**Comment by GawainLynch** [comment]
Ah, I wondered how that got so broad

**Comment by bobdenotter** [comment]
Hey @graham73may, 

This looks like it's good to go, but we're slightly hesitant to merge it into `release/3.2` this close to the 3.2.0 stable. It'll likely go into the 3.2.x releases, after we've done some more monkey-testing. 🐵

**Comment by graham73may** [comment]
Hi @bobdenotter, thanks for the update!

**Comment by GawainLynch** [comment]
Merged into `release/3.2` via 576bfe5a1dc8dd02ff4a591d826850b204320b8a

@graham73may just a note on the merge. I rebased onto  `release/3.2` HEAD to make sure I was getting all changes recently to test … the rebase was kinda "fun" due to the back and forth of changes in the commits to revert the IDE styling, so I have squashed them into one commit … the whole things should be OK, but it just makes for an easier revert in the unlikely event it is needed :smile: 

A very big :+1: to you both on this one!

**Pull request closed without merging**